File: src/emby_range_cache_proxy/origin.py
from __future__ import annotations

import asyncio
from collections.abc import AsyncIterator

from aiohttp import ClientError, ClientSession, ClientTimeout

from .models import ByteRange, SourceMetadata


class OriginError(Exception):
    pass


class OriginClient:
    def __init__(self, *, chunk_bytes: int = 1024 * 1024, timeout_seconds: float = 30.0) -> None:
        self.chunk_bytes = chunk_bytes
        self.timeout_seconds = timeout_seconds
        self._session: ClientSession | None = None

    async def __aenter__(self) -> "OriginClient":
        self._session = ClientSession(
            timeout=ClientTimeout(total=None, sock_connect=self.timeout_seconds, sock_read=self.timeout_seconds)
        )
        return self

    async def __aexit__(self, exc_type, exc, tb) -> None:
        if self._session is not None:
            try:
                await self._session.close()
            finally:
                self._session = None

    async def head(self, url: str) -> SourceMetadata:
        if self._session is None:
            raise RuntimeError("OriginClient must be used as an async context manager")
        try:
            async with self._session.head(url, allow_redirects=True) as response:
                if response.status >= 400:
                    raise OriginError(f"origin HEAD failed: status={response.status}")
                length = _parse_content_length(response.headers.get("Content-Length"))
                return SourceMetadata(
                    url=str(response.url),
                    size=length,
                    etag=response.headers.get("ETag"),
                    last_modified=response.headers.get("Last-Modified"),
                )
        except OriginError:
            raise
        except (asyncio.TimeoutError, TimeoutError):
            raise OriginError("origin HEAD failed: timeout") from None
        except ClientError:
            raise OriginError("origin HEAD failed: client error") from None

    async def stream_range(self, url: str, byte_range: ByteRange) -> AsyncIterator[bytes]:
        if self._session is None:
            raise RuntimeError("OriginClient must be used as an async context manager")
        headers = {"Range": f"bytes={byte_range.start}-{byte_range.end}"}
        try:
            async with self._session.get(url, headers=headers, allow_redirects=True) as response:
                if response.status not in {200, 206}:
                    raise OriginError(f"origin range GET failed: status={response.status}")
                async for chunk in response.content.iter_chunked(self.chunk_bytes):
                    if chunk:
                        yield chunk
        except OriginError:
            raise
        except (asyncio.TimeoutError, TimeoutError):
            raise OriginError("origin range GET failed: timeout") from None
        except ClientError:
            raise OriginError("origin range GET failed: client error") from None


def _parse_content_length(value: str | None) -> int:
    if value is None:
        raise OriginError("origin did not provide Content-Length")
    try:
        length = int(value)
    except ValueError:
        raise OriginError("origin provided invalid Content-Length") from None
    if length < 0:
        raise OriginError("origin provided invalid Content-Length")
    return length

File: go/internal/origin/origin.go
package origin

import (
	"fmt"
	"io"
	"net/http"
	"regexp"
	"strconv"
	"time"

	"github.com/xmm2022/emby-range-cache-proxy/go/internal/model"
)

type Client struct {
	HTTP       *http.Client
	ChunkBytes int64
}

func NewClient(chunkBytes int64) *Client {
	return &Client{
		HTTP: &http.Client{
			Timeout: 0,
			Transport: &http.Transport{
				Proxy:                 http.ProxyFromEnvironment,
				ResponseHeaderTimeout: 30 * time.Second,
			},
		},
		ChunkBytes: chunkBytes,
	}
}

func (c *Client) Head(url string) (model.SourceMetadata, error) {
	req, err := http.NewRequest(http.MethodHead, url, nil)
	if err != nil {
		return model.SourceMetadata{}, err
	}
	resp, err := c.httpClient().Do(req)
	if err != nil {
		return model.SourceMetadata{}, fmt.Errorf("origin HEAD failed: %w", err)
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 400 {
		return model.SourceMetadata{}, fmt.Errorf("origin HEAD failed: status=%d", resp.StatusCode)
	}
	size, err := parseHeadSize(resp)
	if err != nil {
		return model.SourceMetadata{}, err
	}
	return model.SourceMetadata{
		URL:          resp.Request.URL.String(),
		Size:         size,
		ContentType:  resp.Header.Get("Content-Type"),
		ETag:         resp.Header.Get("ETag"),
		LastModified: resp.Header.Get("Last-Modified"),
	}, nil
}

func (c *Client) OpenRange(url string, byteRange model.ByteRange, size int64) (io.ReadCloser, error) {
	req, err := http.NewRequest(http.MethodGet, url, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Range", fmt.Sprintf("bytes=%d-%d", byteRange.Start, byteRange.End))
	resp, err := c.httpClient().Do(req)
	if err != nil {
		return nil, fmt.Errorf("origin range GET failed: %w", err)
	}
	if resp.StatusCode != http.StatusPartialContent {
		resp.Body.Close()
		return nil, fmt.Errorf("origin range GET failed: status=%d", resp.StatusCode)
	}
	if !contentRangeMatches(resp.Header.Get("Content-Range"), byteRange, size) {
		resp.Body.Close()
		return nil, fmt.Errorf("origin range GET failed: invalid Content-Range")
	}
	if length := resp.Header.Get("Content-Length"); length != "" {
		parsed, err := strconv.ParseInt(length, 10, 64)
		if err != nil || parsed != byteRange.Length() {
			resp.Body.Close()
			return nil, fmt.Errorf("origin range GET failed: invalid Content-Length")
		}
	}
	return resp.Body, nil
}

func (c *Client) httpClient() *http.Client {
	if c.HTTP != nil {
		return c.HTTP
	}
	return http.DefaultClient
}

func parseHeadSize(resp *http.Response) (int64, error) {
	if resp.StatusCode == http.StatusPartialContent {
		_, _, total, ok := parseContentRange(resp.Header.Get("Content-Range"))
		if !ok {
			return 0, fmt.Errorf("origin HEAD failed: invalid Content-Range")
		}
		return total, nil
	}
	length := resp.Header.Get("Content-Length")
	if length == "" {
		return 0, fmt.Errorf("origin did not provide Content-Length")
	}
	parsed, err := strconv.ParseInt(length, 10, 64)
	if err != nil || parsed < 0 {
		return 0, fmt.Errorf("origin provided invalid Content-Length")
	}
	return parsed, nil
}

var contentRangeRE = regexp.MustCompile(`^bytes (\d+)-(\d+)/(\d+)$`)

func parseContentRange(value string) (int64, int64, int64, bool) {
	match := contentRangeRE.FindStringSubmatch(value)
	if match == nil {
		return 0, 0, 0, false
	}
	start, err1 := strconv.ParseInt(match[1], 10, 64)
	end, err2 := strconv.ParseInt(match[2], 10, 64)
	total, err3 := strconv.ParseInt(match[3], 10, 64)
	if err1 != nil || err2 != nil || err3 != nil || total <= 0 || start > end || end >= total {
		return 0, 0, 0, false
	}
	return start, end, total, true
}

func contentRangeMatches(value string, byteRange model.ByteRange, size int64) bool {
	start, end, total, ok := parseContentRange(value)
	return ok && start == byteRange.Start && end == byteRange.End && total == size
}
